File: Olympiad/COCI/official/2009/contest5_solutions/lubenica.cpp
#include <algorithm>
#include <iostream>
using namespace std;

#define MAXN 20

int N, H;
int nfriends[MAXN], friendmask[MAXN];

int when[1<<MAXN]; long long thrown[1<<MAXN];

int main() {
   cin >> N >> H;
   for ( int i=0; i<N; ++i ) {
      string row;
      cin >> row;
      for ( int j=0; j<N; ++j ) {
         if ( row[j] == '1' ) {
            ++nfriends[i];
            friendmask[i] |= 1<<j;
         }
      }
   }

   fill(when, when+(1<<N), -1);

   int state = friendmask[0];
   when[state] = 1;
   thrown[state] = nfriends[0];

   long long total = thrown[state];

   for ( int t=2; t<=H; ++t ) {
      int next = 0, lastthrown = 0;
      for ( int i=0; i<N; ++i ) {
         if ( state&(1<<i) ) {
            next ^= friendmask[i];
            lastthrown += nfriends[i];
         } else {
            lastthrown += 2*nfriends[i];
         }
      }
      
      state = next;
      total += lastthrown;

      if ( when[state] == -1 ) {
         when[state] = t;
         thrown[state] = total;
      } else {
         int rep = (H-t) / (t-when[state]);
         t += rep * (t-when[state]);
         total += rep * (total-thrown[state]);
         fill(when, when+(1<<N), -1);
      }
   }
   cout << total << endl;
   return 0;
}

File: Olympiad/COCI/official/2009/contest5_solutions/lubenica.pas
program lubenica;

const MAXN = 20;

var
   N, H, i, j, t, state  : longint;
   next, lastthrown, rep : longint;
   nfriends, friendmask  : array[0..MAXN] of longint;
   when                  : array[0..1 shl MAXN] of longint;
   thrown                : array[0..1 shl MAXN] of int64;
   row                   : string;
   total                 : int64;

begin
   readln(N, H);
   for i:=0 to N-1 do begin
      readln(row);
      for j:=0 to N-1 do begin
         if row[j+1] = '1' then begin
            inc(nfriends[i]);
            friendmask[i] := friendmask[i] or (1 shl j);
         end;
      end;
   end;

   for i:=0 to (1 shl N)-1 do when[i] := -1;

   state := friendmask[0];
   when[state] := 1;
   thrown[state] := nfriends[0];
   total := thrown[state];

   t := 2;
   while t <= H do begin
      next := 0; lastthrown := 0;
      for i:=0 to N-1 do begin
         if state and (1 shl i) <> 0 then begin
            next := next xor friendmask[i];
            lastthrown := lastthrown + nfriends[i];
         end else
            lastthrown := lastthrown + 2*nfriends[i];
      end;
      state := next;
      total := total + lastthrown;

      if when[state] = -1 then begin
         when[state] := t;
         thrown[state] := total;
      end else begin
         rep := (H-t) div (t-when[state]);
         t := t + rep * (t-when[state]);
         total := total + rep * (total-thrown[state]);
         for i:=0 to (1 shl N)-1 do when[i] := -1;
      end;
      inc(t);
   end;
   writeln(total);
end.
